### PR TITLE
fix: gemma3 loss in forward pass

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

Gemma3 in transformers has really high loss due to using old loss calculation.

Tuning on text-only:

image

Tuning on text-vision:

image

`patched` refers to this branch

transformers: `v4.50.3`

Nvidia 4090 `torch 2.6.0+cu124`

### CODE REVIEW

**Comment by NanoCode012** [comment]
Need some eyes on the handling of `attention_mask` , shifting done properly, and whether we need to update `num_items_in_batch`?

**Comment by winglian** [comment]
Screenshot 2025-04-02 at 7 38 38 AM

The fixed train loss and grad norm seems to batch the previous baseline using CCE. So I think this is probably good to go

**Review by winglian — APPROVED** [review]

File: src/axolotl/monkeypatch/gemma3.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        if output_hidden_states is not None
+        else self.config.output_hidden_states
+    )
+    return_dict = (
+        return_dict if return_dict is not None else self.config.use_return_dict
+    )
+
+    is_training = token_type_ids is not None and labels is not None
+
+    # Replace image id woth PAD if the image token if OOV, to avoid index-errors
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/gemma3.py]:
```suggestion
    # Replace image id with PAD if the image token is OOV, to avoid index-errors
```

File: src/axolotl/monkeypatch/gemma3.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    if input_ids is not None and self.config.image_token_index >= self.vocab_size:
+        special_image_mask = input_ids == self.config.image_token_index
+        llm_input_ids = input_ids.clone()
+        llm_input_ids[special_image_mask] = 0
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/gemma3.py:106]: should we grab the pad token here rather than hardcode 0?

**Review by salmanmohammadi — APPROVED** [review]

File: src/axolotl/monkeypatch/gemma3.py (hunk 2 above)
Comment (winglian) [reply src/axolotl/monkeypatch/gemma3.py:106]: looks like this part of the patch is simply duplicated from upstream in `modeling_gemma3.py`
